File: android-native/app/src/main/java/com/reedersystems/commandcomms/signaling/SignalingClient.kt
package com.reedersystems.commandcomms.signaling

import android.util.Log
import io.socket.client.IO
import io.socket.client.Manager
import io.socket.client.Socket
import io.socket.engineio.client.transports.WebSocket
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONObject
import java.util.Timer
import kotlin.concurrent.schedule

private const val TAG = "[PTT-DIAG]"
private const val STARTUP_TAG = "[APP-STARTUP]"

enum class ConnectionState { DISCONNECTED, CONNECTING, CONNECTED, AUTHENTICATED }

class SignalingClient(var serverUrl: String, private var radioToken: String? = null, var deviceId: String? = null) {

    private var socket: Socket? = null

    private val _connectionState = MutableStateFlow(ConnectionState.DISCONNECTED)
    val connectionState: StateFlow<ConnectionState> = _connectionState.asStateFlow()

    private val _events = MutableSharedFlow<SignalingEvent>(extraBufferCapacity = 32)
    val events: SharedFlow<SignalingEvent> = _events.asSharedFlow()

    private val pendingEmergencyEndKeys = mutableSetOf<String>()
    private var authRetryTimer: Timer? = null

    private var unitId: String = ""
    private var username: String = ""

    var onAuthenticated: (() -> Unit)? = null

    fun setRadioToken(token: String?) {
        radioToken = token
    }

    fun connect(unitId: String, username: String) {
        if (_connectionState.value != ConnectionState.DISCONNECTED) return
        this.unitId = unitId
        this.username = username

        Log.d(TAG, "SignalingClient connecting to $serverUrl")
        Log.d(STARTUP_TAG, "SIGNALING_CONNECT_ATTEMPT url=$serverUrl unitId=$unitId")
        _connectionState.value = ConnectionState.CONNECTING

        val optionsBuilder = IO.Options.builder()
            .setPath("/signaling")
            .setTransports(arrayOf(WebSocket.NAME))
            .setReconnection(true)
            .setReconnectionDelay(2_000)
            .setReconnectionAttempts(Integer.MAX_VALUE)

        val token = radioToken
        if (token != null) {
            Log.d(STARTUP_TAG, "SIGNALING_CONNECT_RADIO_AUTH radioToken present")
            optionsBuilder.setQuery("radioToken=$token")
        }

        val options = optionsBuilder.build()
        val s = IO.socket(serverUrl, options)
        socket = s

        s.on(Socket.EVENT_CONNECT) {
            Log.d(TAG, "Socket connected, authenticating as $unitId")
            Log.d(STARTUP_TAG, "SIGNALING_CONNECTED")
            _connectionState.value = ConnectionState.CONNECTED
            val auth = JSONObject().apply {
                put("unitId", unitId)
                put("username", username)
                put("agencyId", "default")
                put("isDispatcher", false)
                deviceId?.let { put("deviceId", it) }
                put("deviceType", if (radioToken != null) "radio" else "desktop")
                put("clientType", "radio")
            }
            Log.d(STARTUP_TAG, "SIGNALING_AUTH_SENT unitId=$unitId deviceId=${deviceId ?: "none"}")
            s.emit("authenticate", auth)

            authRetryTimer?.cancel()
            authRetryTimer = Timer("auth-retry", true).apply {
                schedule(3_000L) {
                    if (_connectionState.value == ConnectionState.CONNECTED) {
                        Log.w(STARTUP_TAG, "SIGNALING_AUTH_RETRY state still CONNECTED after 3s, re-emitting authenticate")
                        s.emit("authenticate", auth)
                    }
                }
            }
        }

        s.on("authenticated") { _ ->
            Log.d(TAG, "Signaling authenticated: $unitId")
            Log.d(STARTUP_TAG, "SIGNALING_AUTH_SUCCESS unitId=$unitId")
            authRetryTimer?.cancel()
            authRetryTimer = null
            _connectionState.value = ConnectionState.AUTHENTICATED
            flushPendingEmergencyEnds()
            onAuthenticated?.invoke()
        }

        s.on("unauthorized") { args ->
            Log.e(STARTUP_TAG, "SIGNALING_AUTH_FAILED reason=${args.firstOrNull()}")
            _connectionState.value = ConnectionState.DISCONNECTED
        }

        s.on("auth:error") { args ->
            Log.e(STARTUP_TAG, "SIGNALING_AUTH_FAILED reason=${args.firstOrNull()}")
            _connectionState.value = ConnectionState.DISCONNECTED
        }

        s.on(Socket.EVENT_DISCONNECT) { _ ->
            Log.d(TAG, "Socket disconnected")
            _connectionState.value = ConnectionState.DISCONNECTED
        }

        s.on(Socket.EVENT_CONNECT_ERROR) { args ->
            Log.w(TAG, "Socket connect error: ${args.firstOrNull()}")
            Log.e(STARTUP_TAG, "SIGNALING_AUTH_FAILED reason=connect_error ${args.firstOrNull()}")
            _connectionState.value = ConnectionState.DISCONNECTED
        }

        s.io().on(Manager.EVENT_RECONNECT_FAILED) {
            Log.e(TAG, "Signaling reconnection attempts exhausted — connection lost")
            _connectionState.value = ConnectionState.DISCONNECTED
        }

        s.on("ptt:pre") { args -> parseAndEmit(args) { json ->
            SignalingEvent.PttPre(
                unitId = json.optString("unitId"),
                channelId = json.optString("channelId")
            )
        }}

        s.on("ptt:start") { args -> parseAndEmit(args) { json ->
            SignalingEvent.PttStart(
                unitId = json.optString("unitId"),
                channelId = json.optString("channelId")
            )
        }}

        s.on("ptt:end") { args -> parseAndEmit(args) { json ->
            SignalingEvent.PttEnd(
                unitId = json.optString("unitId"),
                channelId = json.optString("channelId")
            )
        }}

        s.on("ptt:busy") { args -> parseAndEmit(args) { json ->
            SignalingEvent.PttBusy(
                channelId = json.optString("channelId"),
                transmittingUnit = json.optString("transmittingUnit")
            )
        }}

        s.on("channel:join") { args -> parseAndEmit(args) { json ->
            SignalingEvent.UnitJoined(
                unitId = json.optString("unitId"),
                channelId = json.optString("channelId")
            )
        }}

        s.on("channel:leave") { args -> parseAndEmit(args) { json ->
            SignalingEvent.UnitLeft(
                unitId = json.optString("unitId"),
                channelId = json.optString("channelId")
            )
        }}

        s.on("emergency:start") { args -> parseAndEmit(args) { json ->
            SignalingEvent.EmergencyStart(
                unitId = json.optString("unitId"),
                channelId = json.optString("channelId")
            )
        }}

        s.on("emergency:end") { args -> parseAndEmit(args) { json ->
            SignalingEvent.EmergencyEnd(
                unitId = json.optString("unitId"),
                channelId = json.optString("channelId")
            )
        }}

        s.on("clear_air:start") { args -> parseAndEmit(args) { json ->
            SignalingEvent.ClearAirStart(channelId = json.optString("channelId"))
        }}

        s.on("clear_air:alert") { args -> parseAndEmit(args) { json ->
            SignalingEvent.ClearAirStart(channelId = json.optString("channelId"))
        }}

        s.on("clear_air:end") { args -> parseAndEmit(args) { json ->
            SignalingEvent.ClearAirEnd(channelId = json.optString("channelId"))
        }}

        s.on("unit:status") { args -> parseAndEmit(args) { json ->
            SignalingEvent.UnitStatusChanged(
                unitId = json.optString("unitId"),
                status = json.optString("status")
            )
        }}

        s.on("location:track_start") { args ->
            try {
                val json = args.firstOrNull() as? JSONObject
                val event = SignalingEvent.LocationTrackStart(
                    requestedBy = json?.optString("requestedBy") ?: "dispatch",
                    emergency = json?.optBoolean("emergency", false) ?: false
                )
                _events.tryEmit(event)
            } catch (e: Exception) { Log.w(TAG, "location:track_start parse error") }
        }

        s.on("location:track_stop") { _ ->
            _events.tryEmit(SignalingEvent.LocationTrackStop)
        }

        s.on("radio:channelJoined") { args -> parseAndEmit(args) { json ->
            SignalingEvent.RadioChannelJoined(
                channelId = json.optString("channelId")
            )
        }}

        s.on("ptt:granted") { args -> parseAndEmit(args) { json ->
            val ch = json.optString("channelId")
            val sender = json.optString("senderUnitId")
            Log.d(TAG, "[FloorCtrl] SIGNALING_FLOOR_GRANTED channelId=$ch senderUnitId=$sender")
            SignalingEvent.RadioPttGranted(
                channelId = ch,
                senderUnitId = sender
            )
        }}

        s.on("ptt:denied") { args -> parseAndEmit(args) { json ->
            val ch = json.optString("channelId")
            val reason = json.optString("reason", "")
            val heldBy = json.optString("heldBy")
            Log.d(TAG, "[FloorCtrl] SIGNALING_FLOOR_DENIED channelId=$ch reason=$reason heldBy=$heldBy")
            SignalingEvent.RadioPttDenied(
                channelId = ch,
                reason = reason,
                heldBy = heldBy
            )
        }}

        s.on("tx:start") { args -> parseAndEmit(args) { json ->
            SignalingEvent.RadioTxStart(
                senderUnitId = json.optString("senderUnitId"),
                channelId = json.optString("channelId")
            )
        }}

        s.on("tx:stop") { args -> parseAndEmit(args) { json ->
            SignalingEvent.RadioTxStop(
                senderUnitId = json.optString("senderUnitId"),
                channelId = json.optString("channelId")
            )
        }}

        s.on("channel:busy") { args -> parseAndEmit(args) { json ->
            SignalingEvent.RadioChannelBusy(
                channelId = json.optString("channelId"),
                heldBy = json.optString("heldBy")
            )
        }}

        s.on("channel:floor_taken") { args -> parseAndEmit(args) { json ->
            SignalingEvent.RadioFloorTaken(
                channelId = json.optString("channelId"),
                heldBy = json.optString("heldBy")
            )
        }}

        s.on("channel:idle") { args -> parseAndEmit(args) { json ->
            SignalingEvent.RadioChannelIdle(
                channelId = json.optString("channelId")
            )
        }}

        s.on("radio:dsp_config") { args ->
            try {
                val json = args.firstOrNull() as? JSONObject ?: return@on
                Log.d(TAG, "radio:dsp_config received: $json")
                _events.tryEmit(SignalingEvent.RadioDspConfig(json))
            } catch (e: Exception) {
                Log.w(TAG, "radio:dsp_config parse error: ${e.message}")
            }
        }

        s.on("tx:silence_warning") { args -> parseAndEmit(args) { json ->
            SignalingEvent.TxSilenceWarning(
                unitId = json.optString("unitId"),
                channelId = json.optString("channelId"),
                silenceMs = json.optLong("silenceMs", 0)
            )
        }}

        s.on("radio:locked") { _ ->
            Log.d(TAG, "radio:locked received on main signaling socket")
            _events.tryEmit(SignalingEvent.RadioLocked)
        }

        s.on("radio:unlocked") { _ ->
            Log.d(TAG, "radio:unlocked received on main signaling socket")
            _events.tryEmit(SignalingEvent.RadioUnlocked)
        }

        s.on("radio:unassigned") { _ ->
            Log.d(TAG, "radio:unassigned received on main signaling socket")
            _events.tryEmit(SignalingEvent.RadioUnassigned)
        }

        s.on("radio:kiosk_unlock") { args ->
            try {
                val json = args.firstOrNull() as? JSONObject
                val expiresAt = json?.optLong("expiresAt", 0L) ?: 0L
                val durationMinutes = json?.optInt("durationMinutes", 0) ?: 0
                Log.d(TAG, "radio:kiosk_unlock received expiresAt=$expiresAt durationMinutes=$durationMinutes")
                if (expiresAt > 0L) {
                    _events.tryEmit(SignalingEvent.RemoteKioskUnlock(expiresAt, durationMinutes))
                } else {
                    Log.w(TAG, "radio:kiosk_unlock missing expiresAt — ignoring")
                }
            } catch (e: Exception) {
                Log.w(TAG, "radio:kiosk_unlock parse error: ${e.message}")
            }
        }

        s.on("radio:kiosk_relock") { _ ->
            Log.d(TAG, "radio:kiosk_relock received")
            _events.tryEmit(SignalingEvent.RemoteKioskRelock)
        }

        s.on("radio:assigned") { args ->
            try {
                val json = args.firstOrNull() as? JSONObject
                val assignedUnitId = json?.optString("unitId").orEmpty().trim()
                Log.d(TAG, "radio:assigned received on main signaling socket unitId=$assignedUnitId")
                if (assignedUnitId.isNotBlank()) {
                    _events.tryEmit(SignalingEvent.RadioAssigned(assignedUnitId))
                }
            } catch (e: Exception) {
                Log.w(TAG, "radio:assigned parse error: ${e.message}")
            }
        }

        s.on("ptt:revoked") { args -> parseAndEmit(args) { json ->
            SignalingEvent.PttRevoked(
                channelId = json.optString("channelId"),
                reason = json.optString("reason", "dispatcher_takeover")
            )
        }}

        s.on("ping") { s.emit("pong") }

        s.connect()
    }

    fun disconnect() {
        authRetryTimer?.cancel()
        authRetryTimer = null
        socket?.disconnect()
        socket?.off()
        socket = null
        _connectionState.value = ConnectionState.DISCONNECTED
    }

    fun joinChannel(channelKey: String) {
        if (!isReady()) return
        Log.d(TAG, "joinChannel $channelKey")
        socket?.emit("channel:join", JSONObject().put("channelId", channelKey))
    }

    fun leaveChannel(channelKey: String) {
        if (socket?.connected() != true) return
        Log.d(TAG, "leaveChannel $channelKey")
        socket?.emit("channel:leave", JSONObject().put("channelId", channelKey))
    }

    fun emitPttPre(channelKey: String) {
        if (!isReady()) return
        Log.d(TAG, "emitPttPre $channelKey")
        socket?.emit("ptt:pre", JSONObject().apply {
            put("channelId", channelKey)
            put("unitId", unitId)
        })
    }

    fun emitStatusUpdate(status: String) {
        if (!isReady()) return
        socket?.emit("unit:status", JSONObject().apply {
            put("unitId", unitId)
            put("status", status)
        })
    }

    fun emitEmergencyStart(channelKey: String) {
        if (!isReady()) return
        Log.d(TAG, "emitEmergencyStart $channelKey")
        socket?.emit("emergency:start", JSONObject().put("channelId", channelKey))
    }

    fun emitEmergencyEnd(channelKey: String) {
        if (!isReady()) {
            Log.w(TAG, "emitEmergencyEnd: not authenticated — queuing channelKey=$channelKey for retry on reconnect")
            synchronized(pendingEmergencyEndKeys) {
                pendingEmergencyEndKeys.add(channelKey)
            }
            return
        }
        Log.d(TAG, "emitEmergencyEnd $channelKey")
        synchronized(pendingEmergencyEndKeys) {
            pendingEmergencyEndKeys.remove(channelKey)
        }
        socket?.emit("emergency:end", JSONObject().put("channelId", channelKey))
    }

    fun emitLocationUpdate(lat: Double, lon: Double, accuracy: Float, heading: Float?, speed: Float?) {
        if (socket?.connected() != true) return
        socket?.emit("location:update", JSONObject().apply {
            put("latitude", lat)
            put("longitude", lon)
            put("accuracy", accuracy)
            if (heading != null) put("heading", heading)
            if (speed != null) put("speed", speed)
        })
    }

    fun emitRadioJoinChannel(channelKey: String, udpPort: Int? = null, udpAddress: String? = null) {
        if (!isReady()) return
        Log.d(TAG, "emitRadioJoinChannel $channelKey udpPort=${udpPort ?: "none"} udpAddress=${udpAddress ?: "none"}")
        socket?.emit("radio:joinChannel", JSONObject().apply {
            put("channelId", channelKey)
            if (udpPort != null && udpPort > 0) put("udpPort", udpPort)
            if (!udpAddress.isNullOrBlank() && udpAddress != "0.0.0.0" && udpAddress != "::") put("udpAddress", udpAddress)
        })
    }

    fun emitRadioLeaveChannel(channelKey: String) {
        if (!isReady()) return
        Log.d(TAG, "emitRadioLeaveChannel $channelKey")
        socket?.emit("radio:leaveChannel", JSONObject().put("channelId", channelKey))
    }

    fun emitRadioPttRequest(channelKey: String) {
        if (!isReady()) {
            Log.w(TAG, "[RadioError] emitRadioPttRequest: not ready state=${_connectionState.value} channelKey=$channelKey")
            return
        }
        Log.d(TAG, "[FloorCtrl] SIGNALING_FLOOR_REQUEST channelKey=$channelKey unitId=$unitId sessionTokenPresent=${socket?.connected() == true}")
        socket?.emit("ptt:request", JSONObject().apply {
            put("channelId", channelKey)
            put("unitId", unitId)
        })
    }

    fun emitRadioTxStart(channelKey: String) {
        if (!isReady()) {
            Log.w(TAG, "[RadioError] emitRadioTxStart: not ready state=${_connectionState.value} channelKey=$channelKey")
            return
        }
        Log.d(TAG, "[FloorCtrl] SIGNALING_TX_START channelKey=$channelKey unitId=$unitId")
        socket?.emit("tx:start", JSONObject().apply {
            put("channelId", channelKey)
            put("unitId", unitId)
        })
    }

    fun emitRadioPttRelease(channelKey: String) {
        if (!isReady()) {
            Log.w(TAG, "[RadioError] emitRadioPttRelease: not ready state=${_connectionState.value} channelKey=$channelKey")
            return
        }
        Log.d(TAG, "[FloorCtrl] SIGNALING_FLOOR_RELEASE channelKey=$channelKey unitId=$unitId")
        socket?.emit("ptt:release", JSONObject().apply {
            put("channelId", channelKey)
            put("unitId", unitId)
        })
    }

    fun emitRadioTxStop(channelKey: String) {
        if (!isReady()) {
            Log.w(TAG, "[RadioError] emitRadioTxStop: not ready state=${_connectionState.value} channelKey=$channelKey")
            return
        }
        Log.d(TAG, "[FloorCtrl] SIGNALING_TX_STOP channelKey=$channelKey unitId=$unitId")
        socket?.emit("tx:stop", JSONObject().apply {
            put("channelId", channelKey)
            put("unitId", unitId)
        })
    }

    fun emitRadioSignalQuality(channelKey: String, quality: String, lossPct: Double, jitterMs: Double) {
        if (!isReady()) return
        socket?.emit("radio:signalQuality", JSONObject().apply {
            put("channelId", channelKey)
            put("unitId", unitId)
            put("quality", quality)
            put("lossPct", lossPct)
            put("jitterMs", jitterMs)
            put("timestamp", System.currentTimeMillis())
        })
    }

    fun queryEmergencyStatus(channelKey: String) {
        if (!isReady()) return
        Log.d(TAG, "queryEmergencyStatus channelKey=$channelKey")
        socket?.emit("emergency:status", JSONObject().put("channelId", channelKey),
            io.socket.client.Ack { args ->
                try {
                    val json = args.firstOrNull() as? JSONObject ?: return@Ack
                    val active = json.optBoolean("active", false)
                    val emergencyUnitId = json.optString("unitId", "").ifBlank { null }
                    Log.d(TAG, "emergency:status response channelKey=$channelKey active=$active unitId=$emergencyUnitId")
                    _events.tryEmit(
                        SignalingEvent.EmergencyStatusResponse(
                            channelId = channelKey,
                            active = active,
                            unitId = emergencyUnitId
                        )
                    )
                } catch (e: Exception) {
                    Log.w(TAG, "emergency:status ack parse error: ${e.message}")
                }
            }
        )
    }

    private fun isReady() = _connectionState.value == ConnectionState.AUTHENTICATED

    private fun flushPendingEmergencyEnds() {
        val keys: List<String>
        synchronized(pendingEmergencyEndKeys) {
            keys = pendingEmergencyEndKeys.toList()
        }
        if (keys.isEmpty()) return
        for (key in keys) {
            if (!isReady()) {
                Log.w(TAG, "flushPendingEmergencyEnds: lost auth mid-flush — keeping remaining keys")
                return
            }
            Log.d(TAG, "flushPendingEmergencyEnds: emitting emergency:end for channelKey=$key")
            socket?.emit("emergency:end", JSONObject().put("channelId", key))
            synchronized(pendingEmergencyEndKeys) {
                pendingEmergencyEndKeys.remove(key)
            }
        }
    }

    private inline fun parseAndEmit(
        args: Array<Any>,
        crossinline mapper: (JSONObject) -> SignalingEvent
    ) {
        try {
            val json = args.firstOrNull() as? JSONObject ?: return
            val event = mapper(json)
            _events.tryEmit(event)
        } catch (e: Exception) {
            Log.w(TAG, "parseAndEmit error: ${e.message}")
        }
    }
}

File: ios-native/CommandComms/Networking/SignalingClient.swift
import Foundation
import Combine
import SocketIO

enum SignalingState: String {
    case disconnected
    case connecting
    case connected
    case authenticated
}

enum LocationTrackEvent {
    case start
    case stop
}

enum RadioSignalingEvent {
    case channelJoined(channelId: String, channelIndex: Int)
    case pttGranted(channelId: String, senderUnitId: String)
    case pttDenied(channelId: String, reason: String)
    case pttRevoked(channelId: String, reason: String)
    case txStart(senderUnitId: String, channelId: String)
    case txStop(senderUnitId: String, channelId: String)
    case channelBusy(channelId: String, heldBy: String)
    case channelIdle(channelId: String)
}

@MainActor
final class SignalingClient: ObservableObject {
    @Published var state: SignalingState = .disconnected
    @Published var currentChannel: String?
    @Published var lastError: String?
    @Published var isTransmitting: Bool = false
    @Published var isReceiving: Bool = false
    @Published var isFloorRequestPending: Bool = false
    @Published var lastFloorDenialReason: String?

    let locationTrackEvents = PassthroughSubject<LocationTrackEvent, Never>()

    var serverURL: String
    var sessionCookieHeader: String?

    private var manager: SocketManager?
    private var socket: SocketIOClient?
    private var unitId: String = ""
    private var username: String = ""
    private var pendingChannel: String?

    private let radioSubject = PassthroughSubject<RadioSignalingEvent, Never>()
    var radioEvents: AnyPublisher<RadioSignalingEvent, Never> {
        radioSubject.eraseToAnyPublisher()
    }

    init(serverURL: String) {
        self.serverURL = serverURL
    }

    func updateServerURL(_ url: String) {
        self.serverURL = url
    }

    func connect(unitId: String, username: String, defaultChannel: String?) {
        connect(unitId: unitId, username: username, defaultChannel: defaultChannel, sessionCookie: nil)
    }

    func connect(unitId: String, username: String, defaultChannel: String?, sessionCookie: String?) {
        self.sessionCookieHeader = sessionCookie
        guard state == .disconnected else { return }
        guard let url = URL(string: serverURL) else {
            lastError = "Invalid signaling URL"
            return
        }
        self.unitId = unitId
        self.username = username
        self.pendingChannel = defaultChannel
        self.state = .connecting
        self.lastError = nil

        var config: SocketIOClientConfiguration = [
            .path("/signaling"),
            .forceWebsockets(true),
            .reconnects(true),
            .reconnectWait(2),
            .reconnectAttempts(-1),
            .compress
        ]
        if let cookieHeader = sessionCookieHeader, !cookieHeader.isEmpty {
            config.insert(.extraHeaders(["Cookie": cookieHeader]))
        }
        let manager = SocketManager(socketURL: url, config: config)
        self.manager = manager
        let socket = manager.defaultSocket
        self.socket = socket

        socket.on(clientEvent: .connect) { [weak self] _, _ in
            Task { @MainActor in
                guard let self else { return }
                self.state = .connected
                self.sendAuth()
            }
        }

        socket.on(clientEvent: .disconnect) { [weak self] _, _ in
            Task { @MainActor in
                self?.state = .disconnected
            }
        }

        socket.on(clientEvent: .error) { [weak self] data, _ in
            Task { @MainActor in
                self?.lastError = "\(data.first ?? "connect error")"
            }
        }

        socket.on("authenticated") { [weak self] _, _ in
            Task { @MainActor in
                guard let self else { return }
                self.state = .authenticated
                if let channel = self.pendingChannel, !channel.isEmpty {
                    self.joinChannel(channel)
                }
            }
        }

        socket.on("unauthorized") { [weak self] data, _ in
            Task { @MainActor in
                self?.state = .disconnected
                self?.lastError = "unauthorized: \(data.first ?? "")"
            }
        }

        socket.on("auth:error") { [weak self] data, _ in
            Task { @MainActor in
                self?.state = .disconnected
                self?.lastError = "auth:error: \(data.first ?? "")"
            }
        }

        socket.on("location:track_start") { [weak self] _, _ in
            Task { @MainActor in
                self?.locationTrackEvents.send(.start)
            }
        }

        socket.on("location:track_stop") { [weak self] _, _ in
            Task { @MainActor in
                self?.locationTrackEvents.send(.stop)
            }
        }

        socket.on("channel:join") { [weak self] data, _ in
            Task { @MainActor in
                if let payload = data.first as? [String: Any],
                   let channelId = payload["channelId"] as? String,
                   payload["unitId"] as? String == self?.unitId {
                    self?.currentChannel = channelId
                }
            }
        }

        socket.on("radio:channelJoined") { [weak self] data, _ in
            Task { @MainActor in
                guard let self else { return }
                let payload = data.first as? [String: Any]
                let channel = (payload?["channelId"] as? String) ?? (self.currentChannel ?? "")
                let idxAny = payload?["channelIndex"]
                let channelIndex: Int
                if let n = idxAny as? Int { channelIndex = n }
                else if let n = idxAny as? NSNumber { channelIndex = n.intValue }
                else if let s = idxAny as? String, let n = Int(s) { channelIndex = n }
                else { channelIndex = -1 }
                if !channel.isEmpty {
                    self.currentChannel = channel
                }
                self.radioSubject.send(.channelJoined(channelId: channel, channelIndex: channelIndex))
            }
        }

        socket.on("ptt:granted") { [weak self] data, _ in
            Task { @MainActor in
                guard let self else { return }
                let payload = data.first as? [String: Any]
                let sender = payload?["senderUnitId"] as? String
                let channel = (payload?["channelId"] as? String) ?? (self.currentChannel ?? "")
                self.isFloorRequestPending = false
                self.lastFloorDenialReason = nil
                if sender == nil || sender == self.unitId {
                    self.isTransmitting = true
                }
                self.radioSubject.send(.pttGranted(channelId: channel, senderUnitId: sender ?? self.unitId))
            }
        }

        socket.on("ptt:denied") { [weak self] data, _ in
            Task { @MainActor in
                guard let self else { return }
                let payload = data.first as? [String: Any]
                let channel = (payload?["channelId"] as? String) ?? (self.currentChannel ?? "")
                let reason = (payload?["reason"] as? String) ?? "denied"
                self.isFloorRequestPending = false
                self.isTransmitting = false
                self.lastFloorDenialReason = reason
                self.radioSubject.send(.pttDenied(channelId: channel, reason: reason))
            }
        }

        socket.on("ptt:revoked") { [weak self] data, _ in
            Task { @MainActor in
                guard let self else { return }
                let payload = data.first as? [String: Any]
                let channel = (payload?["channelId"] as? String) ?? (self.currentChannel ?? "")
                let reason = (payload?["reason"] as? String) ?? "dispatcher_takeover"
                self.isFloorRequestPending = false
                self.isTransmitting = false
                self.radioSubject.send(.pttRevoked(channelId: channel, reason: reason))
            }
        }

        socket.on("tx:start") { [weak self] data, _ in
            Task { @MainActor in
                guard let self else { return }
                let payload = data.first as? [String: Any]
                let sender = (payload?["senderUnitId"] as? String) ?? ""
                let channel = (payload?["channelId"] as? String) ?? (self.currentChannel ?? "")
                if sender != self.unitId {
                    self.isReceiving = true
                }
                self.radioSubject.send(.txStart(senderUnitId: sender, channelId: channel))
            }
        }

        socket.on("tx:stop") { [weak self] data, _ in
            Task { @MainActor in
                guard let self else { return }
                let payload = data.first as? [String: Any]
                let sender = (payload?["senderUnitId"] as? String) ?? ""
                let channel = (payload?["channelId"] as? String) ?? (self.currentChannel ?? "")
                if sender == self.unitId {
                    self.isTransmitting = false
                } else {
                    self.isReceiving = false
                }
                self.radioSubject.send(.txStop(senderUnitId: sender, channelId: channel))
            }
        }

        socket.on("channel:floor_taken") { [weak self] data, _ in
            Task { @MainActor in
                guard let self else { return }
                let payload = data.first as? [String: Any]
                let channel = (payload?["channelId"] as? String) ?? (self.currentChannel ?? "")
                let heldBy = (payload?["heldBy"] as? String) ?? ""
                self.isReceiving = true
                self.radioSubject.send(.channelBusy(channelId: channel, heldBy: heldBy))
            }
        }

        socket.on("channel:busy") { [weak self] data, _ in
            Task { @MainActor in
                guard let self else { return }
                let payload = data.first as? [String: Any]
                let channel = (payload?["channelId"] as? String) ?? (self.currentChannel ?? "")
                let heldBy = (payload?["heldBy"] as? String) ?? ""
                self.radioSubject.send(.channelBusy(channelId: channel, heldBy: heldBy))
            }
        }

        socket.on("channel:idle") { [weak self] data, _ in
            Task { @MainActor in
                guard let self else { return }
                let payload = data.first as? [String: Any]
                let channel = (payload?["channelId"] as? String) ?? (self.currentChannel ?? "")
                self.isReceiving = false
                self.radioSubject.send(.channelIdle(channelId: channel))
            }
        }

        socket.connect()
    }

    func joinChannel(_ channelId: String) {
        guard state == .authenticated, let socket else {
            pendingChannel = channelId
            return
        }
        socket.emit("channel:join", ["channelId": channelId])
        socket.emit("radio:joinChannel", ["channelId": channelId])
        currentChannel = channelId
    }

    func leaveChannel(_ channelId: String) {
        guard let socket else { return }
        socket.emit("radio:leaveChannel", ["channelId": channelId])
    }

    func disconnect() {
        socket?.disconnect()
        socket = nil
        manager = nil
        state = .disconnected
        currentChannel = nil
        isTransmitting = false
        isReceiving = false
        isFloorRequestPending = false
    }

    func requestFloor() {
        guard state == .authenticated, let socket, let channel = currentChannel, !channel.isEmpty else {
            lastFloorDenialReason = "not_ready"
            return
        }
        guard !isTransmitting && !isFloorRequestPending else { return }
        isFloorRequestPending = true
        lastFloorDenialReason = nil
        socket.emit("ptt:request", ["channelId": channel, "unitId": unitId])
    }

    func releaseFloor() {
        guard let socket, let channel = currentChannel, !channel.isEmpty else {
            isFloorRequestPending = false
            isTransmitting = false
            return
        }
        let wasActive = isTransmitting || isFloorRequestPending
        isFloorRequestPending = false
        isTransmitting = false
        guard wasActive else { return }
        socket.emit("tx:stop", ["channelId": channel, "unitId": unitId])
        socket.emit("ptt:release", ["channelId": channel, "unitId": unitId])
    }

    func emitLocationUpdate(latitude: Double, longitude: Double, accuracy: Double, heading: Double?, speed: Double?) {
        guard state == .authenticated, let socket else { return }
        var payload: [String: Any] = [
            "latitude": latitude,
            "longitude": longitude,
            "accuracy": accuracy
        ]
        if let heading { payload["heading"] = heading }
        if let speed { payload["speed"] = speed }
        socket.emit("location:update", payload)
    }

    // MARK: - Floor / TX events (used by RadioAudioEngine)

    func emitPttRequest(channelId: String) {
        guard state == .authenticated, let socket else { return }
        isFloorRequestPending = true
        lastFloorDenialReason = nil
        socket.emit("ptt:request", ["channelId": channelId, "unitId": unitId])
    }

    func emitPttRelease(channelId: String) {
        guard state == .authenticated, let socket else { return }
        isFloorRequestPending = false
        isTransmitting = false
        socket.emit("ptt:release", ["channelId": channelId, "unitId": unitId])
    }

    func emitTxStart(channelId: String) {
        guard state == .authenticated, let socket else { return }
        socket.emit("tx:start", ["channelId": channelId, "unitId": unitId])
    }

    func emitTxStop(channelId: String) {
        guard state == .authenticated, let socket else { return }
        socket.emit("tx:stop", ["channelId": channelId, "unitId": unitId])
    }

    func emitRadioSignalQuality(channelId: String, quality: String, lossPct: Double, jitterMs: Double) {
        guard state == .authenticated, let socket else { return }
        let payload: [String: Any] = [
            "channelId": channelId,
            "unitId": unitId,
            "quality": quality,
            "lossPct": lossPct,
            "jitterMs": jitterMs,
            "timestamp": Int(Date().timeIntervalSince1970 * 1000)
        ]
        socket.emit("radio:signalQuality", payload)
    }

    private func sendAuth() {
        guard let socket else { return }
        let payload: [String: Any] = [
            "unitId": unitId,
            "username": username,
            "agencyId": "default",
            "isDispatcher": false,
            "deviceType": "radio",
            "clientType": "radio"
        ]
        socket.emit("authenticate", payload)
    }
}
